#/usr/bin/env bash

rm -r data/py_anno_files
mkdir data/py_anno_files || true

rm -r data/outputs
mkdir data/outputs || true

function annotate() {
	# annotation
	### since I do not understand how to get casette exons from gencode_annotation, it would be an input_data
	python3 annotation.py -gff $1 -o data/py_anno_files/genes.bed
	python3 annotation.py -gff $1 -o data/py_anno_files/exon_gene.tsv | sort -u > data/py_anno_files/exon_gene.tsv
	python3 annotation.py -gff $1 -o data/py_anno_files/stop_codons.bed | intersectBed -b stdin -a data/input_data/cassette_exons.bed -c > data/py_anno_files/stops.tsv
	echo annotation is done
}



function find_cross_targets() {

	factor=$1 # SRSF7 or U2AF1 or all
	KD_p_val_threshold=$2 # p-value threshold for KD of RBPs and NMD KD (typically 1.3 or 1)
	window=$3 # 500 {+-number of nucleotides} around exons where we are looking for peaks (or what you consider appropriate)
	eCLIP_threshold=$4 # threshold for p-value and logFC of eCLIPS (1 or 2)


	# RBPs KD shRNA to reactive exons
	# KD      cell    id      name    deltaPSI        deltaPSIc       z       p       q       KDFC
	# filter not by q-value but by P-VALUE (1)  - consider delta PSI (5th and 6th colums) and only for SRSF7 right now
	awk '$1!=$4' data/input_data/shRNA/B07/$factor*.tsv | grep chr | awk "{ if (\$8 >= $KD_p_val_threshold && (\$6 >= 0.1 ||\$6<=-0.1)) { print } }" > data/outputs/${factor}_reactive_exons.bed
	#awk '$5>0' data/outputs/reactive_exons.bed > data/outputs/reactive_exons_incr.bed
	#awk '$5<0' data/outputs/reactive_exons.bed > data/outputs/reactive_exons_decr.bed

	# eCLIPs

	# instead of using peaks_merged.bed file we start parsing raw_eCLIPs_all.bed (generated by zcat /home/dp/ngs/encodedcc/hg19/eCLIP/*.bed.gz | grep rep) with threshold for -log(p-value) and FDR(?) as 1/1 and 2/2
	awk -v OFS='\t' "{split(\$4,a,\"_\");if(\$7>=$eCLIP_threshold && \$8>=$eCLIP_threshold) print \$1, \$2, \$3, a[1], \$5, \$6, a[2], a[3], \$7, \$8}" data/input_data/raw_eCLIPs_all.bed | sort -u -k1,1 -k2,2n > data/outputs/significant_peaks.bed
    # intersect +-window around exons with significant peaks
	awk -v s=$window '{print $1,$3-s, $4+s, $2,1,$5,$6}' OFS='\t' data/py_anno_files/exon_gene.tsv | bedtools intersect -a stdin -b data/outputs/significant_peaks.bed  -wa -wb -s > data/outputs/exon_peaks.bed
    # keep only cross-interactions
	awk -v OFS='\t' '{if($7!=$11) print $8, $9, $10, $11, $12, $13, $1, $2, $3, $4, $5, $6, $7}' data/outputs/exon_peaks.bed > data/outputs/cross_peaks.bed
	# cross-ine
	python3 filter_cross_regulators2.py data/outputs/${factor}_reactive_exons.bed data/outputs/cross_peaks.bed  > data/outputs/cross_peaks_${factor}.tsv
	python3 filter_by_nmd.py data/input_data/upf1xrn1_deltaPSI.tsv data/outputs/cross_peaks_${factor}.tsv  > data/outputs/cross_peaks_${factor}_nmd.tsv


	# distances between peak and exon? CHECK and change
	#awk -v OFS="\t" '{if($7!=$11){d=(($9+$10)-($2+$3))/2;if(d<0){d=-d};print $4,$2,$3,$7,$11,d}}' data/outputs/exon_peaks.bed > data/outputs/exon_peaks_dist.tsv




}

# find_cross_targets {factor} {window} {eCLIP_threshold} {q_val_threshold}

annotate data/input_data/gencode.v19.annotation.gtf
find_cross_targets SRSF7 1 500 1

